Update forecast.solar sensor values exactly when reported API values change

incaseoftrouble opened the pull request:
## Proposed change

Currently, forecast.solar updates its sensor values every hour, however the API returns values that take effect at a specific time. For example, forecast.solar API may report that at 14:00 the solar power will be 1000W and at 15:00 it will be 2000W. If the forecast.solar integration is loaded at, say, 14:37, it will report 1000W until 15:37 and then update to 2000W for another hour. This change modifies the update interval to always run whenever either the reported value changes or the API can be queried again for new data (respecting the rate limiting).

Marked as WIP since I'm still running longer tests locally (see that it also works correctly around sunset, where forecast.solar gives more frequent updates).

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #72181
- This PR is related to issue: potentially related to #74375 
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

incaseoftrouble committed:
Update sensor values of forecast.solar whenever report API values change + more robust API querying.

homeassistant commented:
Hi @incaseoftrouble, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#77217).

Once you do that we will be able to review and accept this pull request.

Thanks!

incaseoftrouble commented:
Ah this is more annoying than I thought:

If a query to the API happens at, say 14:30, and forecast.solar changes its estimates for 14:00-15:00 then there would be a slight change of the "current hour" values reported at 14:30 even though the values are for 14:00. I'm changing the integration to only change the API values used to compute values at a "change point"

incaseoftrouble committed:
Change API values only whenever a change point is reached (usually hour/half-hour marks and sunrise/sunset)

incaseoftrouble commented:
![image](https://user-images.githubusercontent.com/13201506/186506006-d4fea246-bcd9-41d9-ac0e-e40257967be0.png)

This is how the integration output now looks like. Note how the updates align with the hour marks and sunset.

In a separate PR I could add a "last successful update time" sensor if needed.

incaseoftrouble commented:
Hi @klaasnicolaas any updates?

klaasnicolaas commented on homeassistant/components/forecast_solar/__init__.py:
```diff
@@ -23,6 +34,8 @@
 
 PLATFORMS = [Platform.SENSOR]
 
+_LOGGER = logging.getLogger(__name__)
```
Maybe better to put it in the const.py.

klaasnicolaas reviewed (changes requested):
If you want to use a coordinator I think it's best to do this in a separate file and rename the class to: `ForecastSolarUpdateCoordinator`. Furthermore, I'm not the only one who ultimately has to approve this.

incaseoftrouble commented:
> If you want to use a coordinator I think it's best to do this in a separate file and rename the class to: `ForecastSolarUpdateCoordinator`. Furthermore, I'm not the only one who ultimately has to approve this.

Sorry I was on vacation. Comments are now implemented.

frenck commented on homeassistant/components/forecast_solar/coordinator.py:
```diff
@@ -0,0 +1,163 @@
+            update_method=self.update,
+        )
+        self.api = api
+        # TODO In case there are multiple instances, they should share the api_interval
```
Why is there a TODO here?

incaseoftrouble replied:
Ah, should have clarified this! Because this is a limitation of the forecast solar library. The homeassistant interface to it doesn't / shouldn't know the details of the API limits. This issue was present before already, I just wanted to make it explicit.

The issue is the following: If you create two forecast solar instances (e.g. for a west and east side of a roof) both of them will query once per hour. I am assuming that forecast solar is rate-limiting based on IP, so one IP would effectively query every thirty minutes, exceeding the rate limit.

incaseoftrouble replied:
I reformulated it

frenck reviewed (changes requested):
We should not implement magic and guess towards working with the rate limits. The Forecast.Solar API nicely provides all limits for the used API key in its headers.

incaseoftrouble commented:
Marked this as WIP, I observed some issues I don't understand yet. The sensors sometimes become unavailable in my local deployment.

BTW once this is through I could try to redesign this integration such that you can add multiple planes at once and have the sum of all planes directly available as one sensor (if this is wanted - of course you could derive the sum by creating a sum in the configuration but this might be more convenient?).

frenck commented:
> This was "guessed" before this change already

Incorrect; it used a safe and sane default. It didn't make any approximation.

frenck commented:
> BTW once this is through I could try to redesign this integration such that you can add multiple planes at once and have the sum of all planes directly available as one sensor (if this is wanted - of course you could derive the sum by creating a sum in the configuration but this might be more convenient?).

That seems not to fit in Home Assistants world, as there is no notion of adding multiple things to a single integration. Therefore, this cannot be done without workarounds we probably won't allow.

frenck commented:
Marked the PR as draft and removed "WIP" from the title. As GitHub as native features to mark an PR as in progress/draft.

incaseoftrouble commented:
> Incorrect; it used a safe and sane default. It didn't make any approximation.

```python
    # Free account have a resolution of 1 hour, using that as the default
    # update interval. Using a higher value for accounts with an API key.
    update_interval = timedelta(hours=1)
    if api_key is not None:
        update_interval = timedelta(minutes=30)
```
To me that seems like guessing and does not use the rate limit reported by the API (which actually is 15 min for most models) nor the remaining tries.

Maybe I don't get what you mean by guessing? This code (exactly as the one before) queries the API every `api_interval` time units. The only thing I added to this logic is a) retry more frequently if the connection fails completely (due to `5xx` errors or failure to even establish a connection) and b) delay the querying (if within limits) to get up-to-date data. I don't see a reason not to do (a) and am completely fine with removing (b) - but that has nothing to do with rate limits. If there is a rate limit exception, the integration waits at least `api_interval` until querying again.

On (b): The idea is that querying data at 16:01 is less accurate than querying at 16:59 (since the value changes only at 17:00 anyway). And indeed, the data provided for 17:00 when querying at 16:01 is different from the data provided at 16:59. But again, I am fine with removing that logic, I just thought it would be nice to get more accuracy :-)

> That seems not to fit in Home Assistants world, as there is no notion of adding multiple things to a single integration. Therefore, this cannot be done without workarounds we probably won't allow.

I could imagine having a fixed number of azimuth/declination/kwp fields in the configuration, but I see your point.

> Marked the PR as draft and removed "WIP" from the title. As GitHub has native features to mark an PR as in progress/draft.

Right, sorry, I'm used to GitLab :)

frenck commented:
> To me that seems like guessing and does not use the rate limit reported by the API 

You are missing the point. Anyways, not a relevant discussion. The logic in this PR should not be done.

incaseoftrouble commented:
> You are missing the point.

I'd be happy to understand what it is and fix the problem!

N.B.: I don't mind completely removing any change to how the API is queried (i.e. keep it at a completely fixed interval). That is, indeed, not the point of the PR.

> The logic in this PR should not be done.

Would you maybe explain why? The change significantly improves the quality of the reported values at no cost. Right now the values (and thus the reported curve) can be off by up to an hour. The integration updates the sensor values at "arbitrary" times (`loading time + n * update_interval`), not respecting what the API reports.

incaseoftrouble committed:
Remove randomizing update logic
